### ISSUE TITLE
[C API] Add private "CAST" macros to clean up casts in C code

### ISSUE THREAD

**Issue by vstinner** [python/cpython#91320]
BPO | [47164](https://bugs.python.org/issue47164)
--- | :---
Nosy | @gpshead, @vstinner, @markshannon, @corona10, @Fidget-Spinner
PRs | <li>python/cpython#32190</li><li>python/cpython#32191</li><li>python/cpython#32192</li><li>python/cpython#32210</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-03-30.12:28:10.065>
labels = ['expert-C-API', '3.11']
title = '[C API] Add private "CAST" macros to clean up casts in C code'
updated_at = <Date 2022-04-05.06:59:30.735>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2022-04-05.06:59:30.735>
actor = 'gregory.p.smith'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['C API']
creation = <Date 2022-03-30.12:28:10.065>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 47164
keywords = ['patch']
message_count = 14.0
messages = ['416350', '416353', '416354', '416357', '416358', '416361', '416362', '416363', '416365', '416372', '416417', '416418', '416419', '416420']
nosy_count = 5.0
nosy_names = ['gregory.p.smith', 'vstinner', 'Mark.Shannon', 'corona10', 'kj']
pr_nums = ['32190', '32191', '32192', '32210']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue47164'
versions = ['Python 3.11']
```

</p></details>

**Comment by vstinner** [comment]
Concrete problem of METH_NOARG undefined behavior on a function with a single parameter, whereas the ABI requires 2 parameters (pass NULL as the second parameter): https://blog.pyodide.org/posts/function-pointer-cast-handling/

It's also an issue for CFI.

**Comment by vstinner** [comment]
CAST macros now use reinterpret_cast<> on C++: https://github.com/python/cpython/commit/29e2245ad5d28b565fe0d0d667d283708c6e832c of #91959. I close the issue.

**Comment by vstinner** [comment]
Hum, the issue was not fully fixed (especially, related to C++: see issue #91321). I pushed a few more changes:

* commit b270b82f1137ff25ee263eafd31503d760f3403d: Argument Clinic uses _PyCFunction_CAST() (#32210)
* commit ff3e9cdf334737aeedbbec2e5a219084d27db9cd: Fix more old-style cast warnings in C++ (#92247)
* commit 804f2529d8e545a8d59eaf260ee032d014e681ba: Use _PyCFunction_CAST() (#92251)
